pallets/process-validation/src/lib.rs

```diff
@@ -90,38 +92,140 @@ pub mod pallet {
     }
 
     #[pallet::error]
     pub enum Error<T> {
-        // TODO: implement errors for extrinsics
+        // attempting to disable non-existing process
+        NonExistingProcess,
+        // process is already disabled
+        AlreadyDisabled,
+        // process not found for this versiion
+        InvalidVersion,
     }
 
     // The pallet's dispatchable functions.
     #[pallet::call]
     impl<T: Config> Pallet<T> {
-        // TODO: implement create_process with correct parameters and impl
         #[pallet::weight(T::WeightInfo::create_process())]
-        pub(super) fn create_process(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
-            // Check it was signed and get the signer
+        pub(super) fn create_process(
+            origin: OriginFor<T>,
+            id: T::ProcessIdentifier,
+            restrictions: Vec<Restriction>,
+        ) -> DispatchResultWithPostInfo {
             T::CreateProcessOrigin::ensure_origin(origin)?;
+            let new_version: T::ProcessVersion = Pallet::<T>::update_version(id.clone()).unwrap();
+            Pallet::<T>::persist_process(&id, &new_version, restrictions.clone());
+
+            Self::deposit_event(Event::ProcessCreated(
+                id,
+                new_version.clone(),
+                restrictions,
+                new_version == One::one(),
+            ));
 
-            Self::deposit_event(Event::ProcessCreated);
-            Ok(().into())
+            return Ok(().into());
         }
 
-        // TODO: implement disable_process with correct parameters and impl
         #[pallet::weight(T::WeightInfo::disable_process())]
-        pub(super) fn disable_process(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+        pub(super) fn disable_process(
+            origin: OriginFor<T>,
+            id: T::ProcessIdentifier,
+            version: T::ProcessVersion,
+        ) -> DispatchResultWithPostInfo {
             T::DisableProcessOrigin::ensure_origin(origin)?;
-            Self::deposit_event(Event::ProcessDisabled);
-            Ok(().into())
+
+            Pallet::<T>::validate_version_and_id(&id, &version)?;
+            Pallet::<T>::set_disabled(&id, &version)?;
+
+            Self::deposit_event(Event::ProcessDisabled(id, version));
+            return Ok(().into());
+        }
+    }
+
+    // helper methods
+    impl<T: Config> Pallet<T> {
+        pub fn get_version(id: &T::ProcessIdentifier) -> T::ProcessVersion {
+            let version: T::ProcessVersion = <VersionModel<T>>::get(&id);
+            return version;
+        }
+
+        // rebase with master
+        pub fn update_version(id: T::ProcessIdentifier) -> Result<T::ProcessVersion, ()> {
+            let version: T::ProcessVersion = Pallet::<T>::get_version(&id);
+            let exists: bool = <ProcessModel<T>>::contains_key(&id, version.clone());
+            let new_version: T::ProcessVersion = match exists {
```

Comment:
The logic here is dodgey. I don't think `exists` is ever `true`. If `id` doesn't exist `version` will be the default value of `T::ProcessVersion` which for our impl is `0`. That will make `version` `1` which is what we say is new. This should definitely be cleaned up

Reply:
Thanks,. going to back this up with the new test scenario

Reply:
This has been updated, going to mark as resolved